### PR TITLE
[Feature] Nested tensordicts

### PR DESCRIPTION
**Pull request by vmoens**
- [X] tests
- [x] flatten (do we need that?)  => #264 
- [x] nested composite spec -> test rand and encode => #265 
- [x] indexing using tuples `td[key1, key2]" => #262

```python
from torchrl.data import TensorDict
import torch

td1 = TensorDict({"a": torch.randn(4, 1)}, [4])
td2 = TensorDict({"b": torch.randn(4, 1), "td": td1}, [4])

print(td2)

td1 = TensorDict({"a": torch.randn(4, 1)}, [])
td2 = TensorDict({"b": torch.randn(4, 1), "td": td1}, [4])

print(td2)

td1 = TensorDict({"a": torch.randn(4, 1)}, [])
td2 = TensorDict({"b": torch.randn(4, 1)}, [4])
td2["td"] = td1

print(td2)

td1 = TensorDict({"a": torch.randn(4, 1)}, [])
td2b = TensorDict({"b": torch.randn(4, 1)}, [4])
td2b["td"] = td1

td2b[:] = td2

print(td2)

```

### CODE REVIEW

File: torchrl/data/tensordict/tensordict.py [hunk 1]
```diff
@@ -67,6 +66,7 @@ class _TensorDict(Mapping, metaclass=abc.ABCMeta):
 
     _safe = False
     _lazy = False
+    is_meta = False
```
Comment (shagunsodhani) [inline torchrl/data/tensordict/tensordict.py:69]: not a big deal but you want to name this as `_meta` to be consistent with the rest of the code?

File: torchrl/data/tensordict/metatensor.py [hunk 2]
```diff
@@ -70,7 +70,8 @@ def __init__(
         _is_shared: Optional[bool] = None,
         _is_memmap: Optional[bool] = None,
     ):
-
+        _is_tensordict = False
+        _repr = None
```
Comment (shagunsodhani) [inline torchrl/data/tensordict/metatensor.py:74]: If we want `repr` to be non-None only for tensordict, we may rename this variable to be something else?

File: torchrl/data/tensordict/tensordict.py (hunk 1 above)
Comment (vmoens) [reply torchrl/data/tensordict/tensordict.py:69]: it comes from the fact that we have checks whether a tensor has [an attribute "is_meta"](https://pytorch.org/docs/stable/generated/torch.Tensor.is_meta.html) (i.e. is it on "meta" device)
TensorDict does not support meta device...

File: torchrl/data/tensordict/metatensor.py (hunk 2 above)
Comment (vmoens) [reply torchrl/data/tensordict/metatensor.py:74]: refactored! Thanks for the suggestion

**Pull request merged**